Profile creation

ermac95 opened the pull request:
- Created Profile, ProfileEdit and BecomeSpeaker Fragments
- Created ProfileEditing system
- Created setting avatar with gallery/camera picture
- Saving profile data in SharedPreference

ermac95 committed:
- Created Profile, ProfileEdit and BecomeSpeaker Fragments
- Created ProfileEditing system
- Created setting avatar with gallery/camera picture
- Saving profile data in SharedPreference

ermolnik commented on app/src/main/java/com/academy/android/MainActivity.kt:
```diff
@@ -12,12 +14,14 @@ import dagger.hilt.android.AndroidEntryPoint
 @AndroidEntryPoint
 class MainActivity : AppCompatActivity() {
 
+    private lateinit var navController: NavController
```
я сторонник того, чтобы везде, где это возможно использовать Null-able поля вместо lateinit. Lateinit использовать только для Inject-ов. Поскольку с lateinit намного проще ошибиться и вызывать его до момента, когда он будет инициализирван. Здесь не важно, можем оставить так, пишу просто так для размышления

ermac95 replied:
Оставил как lateinit, потому что при установке контроллера не позволяет предать изменяемую переменную, а проверка if (navController != null) его не устраивает, чтобы не ставить navController!!, что вроде как из курса академии нежелательно использовать, я оставлю его тогда как lateinit, он тут один тем более. В остальных местах больше lateinit переменных нет.

AlekseyYakovlev replied:
В данном случае лучше использовать иммутабельную переменную с ленивой инициализацией.
Например, так:
private val navController: NavController by lazy { findNavController(R.id.nav_host_fragment) }

ermac95 replied:
Кстати да, забыл про этот метод, поправил, запушил

ermolnik commented on app/src/main/java/com/academy/android/ui/profile/EditProfileFragment.kt:
```diff
@@ -0,0 +1,170 @@
+        vb.btnSaveProfile.setOnClickListener {
+            when {
+                TextUtils.isEmpty(vb.fieldEditName.text) -> {
+                    vb.fieldEditName.error = "Похоже, вы забыли заполнить поле \"Ваше имя\""
```
хорошей практикой является не хардкодить стринги в фрагменте а выносить их в strings.xml, поправишь?

ermac95 replied:
Поправил

ermolnik commented on app/src/main/java/com/academy/android/ui/profile/ProfileFragment.kt:
```diff
@@ -1,32 +1,50 @@
-                vb.textProfile.text = it
+            profileViewModel.profileData.collectLatest { profile ->
+                vb.editProfileCard.load(profile.profPic)
+                //vb.editProfileCard.setImageURI(profileUri)
```
Не нужно оставлять закомментированного кода в деве, убери пожалуйста

ermac95 replied:
Убрал

ermolnik commented on app/src/main/res/drawable/profile_field_bg.xml:
```diff
@@ -0,0 +1,8 @@
+    android:shape="rectangle">
+
+    <solid android:color="@color/light_green" />
+    <corners android:radius="15dp"/>
```
обычно все скругления и отступы должны быть кратные 2. Поправь на 16

ermac95 replied:
Поправил

ermac95 committed:
- Fixed all the review's remarks and comments

ermac95 committed:
- Fixed initialization of navController with lazy

AlekseyYakovlev commented on app/build.gradle.kts:
```diff
@@ -5,6 +5,7 @@ plugins {
     kotlin("kapt")
     id("dagger.hilt.android.plugin")
     id("androidx.navigation.safeargs.kotlin")
+    id("kotlin-android")
```
Зачем второй раз добавлять плагин "kotlin-android"?
При использовании kts android плагин мы добавляем через kotlin("android")

ermac95 replied:
Видимо когда пересобирал проект из за косяка с новой либой, он мне автоматом создал, убрал.

AlekseyYakovlev commented on app/build.gradle.kts:
```diff
@@ -66,6 +67,8 @@ dependencies {
     // Navigation
     implementation("androidx.navigation:navigation-fragment-ktx:${rootProject.extra["navigationVersion"]}")
     implementation("androidx.navigation:navigation-ui-ktx:${rootProject.extra["navigationVersion"]}")
+    implementation("org.jetbrains.kotlin:kotlin-stdlib:${rootProject.extra["kotlin_version"]}")
```
Стандартная библиотека котлин добавляется в проект по умолчанию, прописывать её в импорты не нужно.

ermac95 replied:
Навигацию вроде как не трогал, не знаю откуда она там, опять же возможно из-за пересборки, убрал

AlekseyYakovlev commented on app/build.gradle.kts:
```diff
@@ -66,6 +67,8 @@ dependencies {
     implementation("androidx.navigation:navigation-fragment-ktx:${rootProject.extra["navigationVersion"]}")
     implementation("androidx.navigation:navigation-ui-ktx:${rootProject.extra["navigationVersion"]}")
+    implementation("org.jetbrains.kotlin:kotlin-stdlib:${rootProject.extra["kotlin_version"]}")
+    implementation("androidx.legacy:legacy-support-v4:1.0.0")
```
Для чего используется данная библиотека? Если не используется, то стоит убрать.

ermac95 replied:
Убрал

AlekseyYakovlev commented on app/build.gradle.kts:
```diff
@@ -95,6 +98,12 @@ dependencies {
+    implementation ("androidx.preference:preference-ktx:1.1.1")
+
+    //Gson
+    implementation ("com.google.code.gson:gson:2.8.6")
```
Для сериализации данных в проекте ранее договаривались использовать kotlinx-serialization. Не стоит дополнительно тащить в проект Gson.

ermac95 replied:
Заменил

AlekseyYakovlev commented on app/src/main/res/values-ru-rRU/strings.xml:
```diff
@@ -6,4 +6,29 @@
+    <string name="edit_name">Ваше имя</string>
+    <string name="edit_surname">Ваша фамилия</string>
+    <string name="edit_number">Телефон</string>
+    <string name="edit_email">Почтовый адрес</string>
```
Подсказка "Почтовый адрес" путает пользователя. 
Лучше написать "Адрес электронной почты" или просто "Email".

ermac95 replied:
Заменил

AlekseyYakovlev commented on app/src/main/res/layout/fragment_profile.xml:
```diff
@@ -1,22 +1,237 @@
 <?xml version="1.0" encoding="utf-8"?>
-<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+
+<androidx.constraintlayout.widget.ConstraintLayout
```
Стоит подумать над следующими моментами в вёрстке:
Кнопка "Сохранить" перекрывает часть полей.
Отсутствует плейсхолдер для аватара.
Значок editProfileCard частично перекрывает аватар пользователя. Я предлагаю уменьшить его или скрывать если аватар не дефолтный, а действие по смене аватара перепривязать к клику по самому аватару.  

![image](https://user-images.githubusercontent.com/31004501/115154626-91f7de80-a084-11eb-8e91-0c2442994717.png)

ermac95 committed:
- Fixed all the issues mentioned in review

ermac95 committed:
- Implemented button animation fade_out upon scrolling down and fade upon scrolling up
- Fixed some UI Profile and EditProfile issues
- Added avatar placeholder and decreased size of change ava button

AlekseyYakovlev commented on app/src/main/java/com/academy/android/ui/profile/EditProfileFragment.kt:
```diff
@@ -0,0 +1,170 @@
+        }
+    }
+
+    override fun onCreateView(
```
Можно в конструкторе фрагмента просто указать файл разметки 
class EditProfileFragment: Fragment(R.layout.fragment_edit_profile) 
и тогда onCreateView переопределять не нужно.

AlekseyYakovlev commented on app/src/main/java/com/academy/android/ui/profile/ImagePickerDialog.kt:
```diff
@@ -0,0 +1,46 @@
+
+    private val vb by viewBinding(AddProfPicDialogBinding::bind)
+
+    override fun onCreateView(
```
Можно в конструкторе диалога просто указать файл разметки и тогда onCreateView переопределять не нужно.

AlekseyYakovlev commented on app/src/main/java/com/academy/android/ui/profile/ProfBecomeSpeakerFragment.kt:
```diff
@@ -0,0 +1,36 @@
+
+    private val vb by viewBinding(FragmentProfBecomeSpeakerBinding::bind)
+
+    override fun onCreateView(
```
Можно в конструкторе фрагмента просто указать файл разметки и тогда onCreateView переопределять не нужно.

AlekseyYakovlev commented on app/src/main/res/drawable/button_bg_style.xml:
```diff
@@ -0,0 +1,8 @@
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+
+    <solid android:color="#27C12E" />
```
Цвета лучше брать из темы #27C12E будет ?attr/colorPrimary

AlekseyYakovlev commented on app/src/main/res/drawable/button_bg_style.xml:
```diff
@@ -0,0 +1,8 @@
+    android:shape="rectangle">
+
+    <solid android:color="#27C12E" />
+    <corners android:radius="50dp"/>
```
Размеры лучше вынести в dimens.xml

AlekseyYakovlev commented on app/src/main/res/drawable/ic_add_picture.xml:
```diff
@@ -0,0 +1,10 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
```
Размеры лучше вынести в dimens.xml

AlekseyYakovlev commented on app/src/main/res/drawable/ic_arrow_back.xml:
```diff
@@ -0,0 +1,11 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
```
Размеры лучше вынести в dimens.xml

AlekseyYakovlev commented on app/src/main/res/drawable/ic_arrow_forward.xml:
```diff
@@ -0,0 +1,11 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
```
Размеры лучше вынести в dimens.xml

AlekseyYakovlev commented on app/src/main/res/drawable/ic_camera_picker.xml:
```diff
@@ -0,0 +1,13 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
```
Размеры лучше вынести в dimens.xml

AlekseyYakovlev commented on app/src/main/res/drawable/ic_gallery_picker.xml:
```diff
@@ -0,0 +1,10 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
```
Размеры лучше вынести в dimens.xml

AlekseyYakovlev commented on app/src/main/res/layout/add_prof_pic_dialog.xml:
```diff
@@ -0,0 +1,76 @@
+        android:layout_height="wrap_content"
+        android:layout_margin="10dp"
+        android:layout_weight="1"
+        app:cardCornerRadius="20dp"
```
Размеры элементов и радиусы скруглений лучше вынести в dimens.xml

AlekseyYakovlev commented on app/src/main/res/layout/add_prof_pic_dialog.xml:
```diff
@@ -0,0 +1,76 @@
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            android:orientation="vertical"
+            android:background="@color/light_green"
```
Цвета лучше не использовать напрямую, а создать и использовать новый цвет в теме или переиспользовать существующий из незадействованных

AlekseyYakovlev commented on app/src/main/res/layout/fragment_edit_profile.xml:
```diff
@@ -0,0 +1,169 @@
+
+    </ScrollView>
+
+    <Button
```
Стоит подумать над следующими моментами в вёрстке:
Кнопка "Сохранить" перекрывает часть полей.
Отсутствует плейсхолдер для аватара. 

![image](https://user-images.githubusercontent.com/31004501/115155210-8a860480-a087-11eb-91e8-db3a811ba2ce.png)

AlekseyYakovlev commented on app/src/main/res/layout/fragment_profile.xml:
```diff
@@ -1,22 +1,237 @@
+        app:layout_constraintBottom_toTopOf="@id/main_guideline"
+        app:layout_constraintEnd_toEndOf="parent" />
+
+    <androidx.cardview.widget.CardView
```
В данном случае лучше просто использовать ShapeableImageView

AlekseyYakovlev commented on app/src/main/res/layout/fragment_profile.xml:
```diff
@@ -1,22 +1,237 @@
+        app:layout_constraintTop_toBottomOf="@id/profileCard"
+        tools:text="Student" />
+
+    <ScrollView
```
У элементов внутри ScrollView обрезаются тени
![image](https://user-images.githubusercontent.com/31004501/115470226-a67fd680-a23e-11eb-861b-52ebbe56e21e.png)

ermac95 committed:
- Changed ProfileAvatar from CardView to ShapeableImageView
- Transferred general dimensions to dimens file
- Fixed small bugs, mentioned in review

ermac95 committed:
Merge remote-tracking branch 'origin/develop' into feature/profile

# Conflicts:
#	app/src/main/java/com/academy/android/di/RepositoriesModule.kt
#	app/src/main/res/values-ru-rRU/strings.xml
#	app/src/main/res/values/strings.xml

ermac95 committed:
Merge remote-tracking branch 'origin/develop' into feature/profile

# Conflicts:
#	app/src/main/java/com/academy/android/di/RepositoriesModule.kt
#	app/src/main/res/values-ru-rRU/strings.xml
#	app/src/main/res/values/strings.xml

ermac95 committed:
Merge remote-tracking branch 'origin/feature/profile' into feature/profile
